File: AKManagedObject.m
//
//  AKManagedObject.m
//  AKCoreDataController
//
//  Created by Ken M. Haggerty on 4/13/16.
//  Copyright © 2016 Flatiron School. All rights reserved.
//

#pragma mark - // NOTES (Private) //

#pragma mark - // IMPORTS (Private) //

#import "AKManagedObject.h"
//#import "AKDebugger.h"
#import "AKGenerics.h"

#pragma mark - // DEFINITIONS (Private) //

NSString * const AKManagedObjectWillBeDeallocatedNotification = @"kNotificationAKManagedObject_WillBeDeallocated";
NSString * const AKManagedObjectWasCreatedNotification = @"kNotificationAKManagedObject_WasCreated";
NSString * const AKManagedObjectWasFetchedNotification = @"kNotificationAKManagedObject_WasFetched";
NSString * const AKManagedObjectWillSaveNotification = @"kNotificationAKManagedObject_WillSave";
NSString * const AKManagedObjectDidSaveNotification = @"kNotificationAKManagedObject_DidSave";
NSString * const AKManagedObjectWillBeDeletedNotification = @"kNotificationAKManagedObject_WillBeDeleted";

@interface AKManagedObject ()
@property (nonatomic, strong, readwrite, nullable) NSSet *changedKeys;
@property (nonatomic, readwrite) BOOL isSaving;
@property (nonatomic, readwrite) BOOL willBeDeleted;
@property (nonatomic, readwrite) BOOL wasDeleted;
@end

@implementation AKManagedObject

#pragma mark - // SETTERS AND GETTERS //

@synthesize changedKeys = _changedKeys;
@synthesize isSaving = _isSaving;
@synthesize willBeDeleted = _willBeDeleted;
@synthesize wasDeleted = _wasDeleted;
@synthesize parentIsDeleted = _parentIsDeleted;

#pragma mark - // INITS AND LOADS //

- (void)dealloc {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeSetup tags:@[AKD_CORE_DATA] message:nil];
    
    [NSNotificationCenter postNotificationToMainThread:AKManagedObjectWillBeDeallocatedNotification object:self userInfo:nil];
}

- (void)awakeFromInsert {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeSetup tags:@[AKD_CORE_DATA] message:nil];
    
    [super awakeFromInsert];
    
    [NSNotificationCenter postNotificationToMainThread:AKManagedObjectWasCreatedNotification object:self userInfo:nil];
}

- (void)awakeFromFetch {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeSetup tags:@[AKD_CORE_DATA] message:nil];
    
    [super awakeFromFetch];
    
    [NSNotificationCenter postNotificationToMainThread:AKManagedObjectWasFetchedNotification object:self userInfo:nil];
}

- (void)willSave {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeSetup tags:@[AKD_CORE_DATA] message:nil];
    
    self.isSaving = YES;
    
    [super willSave];
    
    if (self.isUpdated && !self.isInserted) {
        self.changedKeys = [NSMutableSet setWithArray:self.changedValues.allKeys];
    }
    
    NSDictionary *userInfo = [NSDictionary dictionaryWithNullableObject:self.changedKeys forKey:NOTIFICATION_OBJECT_KEY];
    [NSNotificationCenter postNotificationToMainThread:AKManagedObjectWillSaveNotification object:self userInfo:userInfo];
}

- (void)didSave {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeSetup tags:@[AKD_CORE_DATA] message:nil];
    
    if (self.willBeDeleted) {
        self.willBeDeleted = NO;
        self.wasDeleted = YES;
    }
    
    if (self.changedKeys && !self.inserted) { // !self.isDeleted &&
        [NSNotificationCenter postNotificationToMainThread:AKManagedObjectDidSaveNotification object:self userInfo:@{NOTIFICATION_OBJECT_KEY : self.changedKeys}];
        self.changedKeys = nil;
    }
    
    [super didSave];
    
    self.isSaving = NO;
}

- (void)prepareForDeletion {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeUnspecified tags:@[AKD_CORE_DATA] message:nil];
    
    self.willBeDeleted = YES;
    
    [NSNotificationCenter postNotificationToMainThread:AKManagedObjectWillBeDeletedNotification object:self userInfo:nil];
    
    [super prepareForDeletion];
}

#pragma mark - // PUBLIC METHODS //

#pragma mark - // CATEGORY METHODS //

#pragma mark - // DELEGATED METHODS //

#pragma mark - // OVERWRITTEN METHODS //

- (void)setup {
//    [AKDebugger logMethod:METHOD_NAME logType:AKLogTypeMethodName methodType:AKMethodTypeSetup tags:@[AKD_CORE_DATA] message:nil];
    
    [super setup];
    
    self.isSaving = NO;
    self.wasDeleted = NO;
    self.parentIsDeleted = NO;
}

#pragma mark - // PRIVATE METHODS //

@end

File: AKManagedObject.swift
//
//  AKManagedObject.swift
//  pcyp-ios
//
//  Created by Ken M. Haggerty on 6/14/16.
//  Copyright © 2016 Peter Cicchino Youth Project. All rights reserved.
//

// MARK: Imports

import Foundation
import CoreData

// MARK: Protocols

// MARK: Definitions

let AKManagedObjectWillBeDeallocatedNotification = "kAKManagedObjectWillBeDeallocatedNotification"
let AKManagedObjectWasCreatedNotification = "kAKManagedObjectWasCreatedNotification"
let AKManagedObjectWasFetchedNotification = "kAKManagedObjectWasFetchedNotification"
let AKManagedObjectWillSaveNotification = "kAKManagedObjectWillSaveNotification"
let AKManagedObjectDidSaveNotification = "kAKManagedObjectDidSaveNotification"
let AKManagedObjectWillBeDeletedNotification = "kAKManagedObjectWillBeDeletedNotification"

// MARK: - AKManagedObject

class AKManagedObject: NSManagedObject {
    
    // MARK: - Public API
    
    // MARK: • IBOutlets
    
    // MARK: • Variables
    
    private(set) var changedKeys: [String]?
    private(set) var isSaving: Bool = false
    private(set) var willBeDeleted: Bool = false
    private(set) var wasDeleted: Bool = false
    private(set) var parentIsDeleted: Bool = false
    
    // MARK: • Functions
    
    // MARK: - Private API
    
    // MARK: • IBOutlets
    
    // MARK: • Variables
    
    // MARK: • Inits
    
    deinit {
        NSNotificationCenter.postNotificationToMainThread(AKManagedObjectWillBeDeallocatedNotification, object: self, userInfo: nil)
    }
    
    override init(entity: NSEntityDescription, insertIntoManagedObjectContext context: NSManagedObjectContext?) {
        super.init(entity: entity, insertIntoManagedObjectContext: context)
    }
    
    override func awakeFromInsert() {
        super.awakeFromInsert()
        
        NSNotificationCenter.postNotificationToMainThread(AKManagedObjectWasCreatedNotification, object: self, userInfo: nil)
    }
    
    override func awakeFromFetch() {
        super.awakeFromFetch()
        
        NSNotificationCenter.postNotificationToMainThread(AKManagedObjectWasFetchedNotification, object: self, userInfo: nil)
    }
    
    override func willSave() {
        self.isSaving = true
        
        super.willSave()
        
        if (self.updated && !self.inserted) {
            self.changedKeys = Array(self.changedValues().keys)
        }
        
        NSNotificationCenter.postNotificationToMainThread(AKManagedObjectWillSaveNotification, object: self, userInfo: [NOTIFICATION_OBJECT_KEY : self.changedKeys])
    }
    
    override func didSave() {
        if (self.willBeDeleted) {
            self.willBeDeleted = false
            self.wasDeleted = true
        }
        
        if ((self.changedKeys != nil) && !self.inserted) { // && !self.isDeleted
            NSNotificationCenter.postNotificationToMainThread(AKManagedObjectDidSaveNotification, object: self, userInfo: [NOTIFICATION_OBJECT_KEY : self.changedKeys!])
            self.changedKeys = nil
        }
        
        super.didSave()
        
        self.isSaving = false
    }
    
    // MARK: • Setup
    
    // MARK: • Delegated
    
    // MARK: • Functions

}
